Add an option sieve for all commands

gigabo opened the pull request:
Allow config:
- At the top level of `asini.json`
- In `commands.<command>` in `asini.json`
- On the command-line

Each of those supercedes previous values.

Example:

``` json
{
  "asini": "x.x.x",
  "version": "1.2.0",
  "ignore": "unrelated-package",
  "commands": {
    "publish": {
      "ignore": "test-package"
    }
  }
}
```

This deprecates:
- `bootstrapConfig.ignore`
- `publishConfig.ignore`

These still work, with the _lowest_ precedence.  Their use triggers a warning.

This addresses some of the TODOs and feedback from #8.

gigabo committed:
Add an option sieve for all commands

Allow config:

- At the top level of `asini.json`
- In `commands.<command>` in `asini.json`
- On the command-line

Each of those supercedes previous values.

Example:

```json
{
  "asini": "x.x.x",
  "version": "1.2.0",
  "ignore": "unrelated-package",
  "commands": {
    "publish": {
      "ignore": "test-package"
    }
  }
}
```

This deprecates:

- `bootstrapConfig.ignore`
- `publishConfig.ignore`

These still work, with the _lowest_ precedence.  Their use triggers a warning.

gigabo commented:
Ugh... no `Object.assign` in node 0.10 or 0.12...

Maybe we can just drop those?

Since we're not required to support everything that babel supports now? 👹

rygine commented on src/Command.js:
```diff
@@ -19,6 +19,47 @@ export default class Command {
+  }
+
+  get className() {
+    return this.constructor.name;
```
not a big fan of this, but i guess it works. ¯_(ツ)_/¯

gigabo replied:
You would prefer each command implement a name getter?

Like:

``` javascript
class LsCommand extends Command {
  ...
  get name() { return "ls" }
}
```

That's more explicit and flexible... but it's a lot of boilerplate, too.

rygine replied:
something like that. yes, a bit more boilerplate, but would support classenames without `Command` in them and additional characters. not sure there's an advantage there, just a thought.

rygine replied:
and in the future, when `asini` is an `asini` repo, custom commands should be possible, and names could _potentially_ clash. it might be nice to be explicit with the name rather than relying on the class name.

gigabo replied:
Yeah, I see your point.  Feels a little bit magic to base it on the class name.

My thought is that it's easier to get `class *Command` right than it is to set up a name getter.  It also doesn't seem very onerous to require a naming convention for command classes.  They will be relatively few, and they operate in the context of a framework.

How about I add a check for the `Command` suffix on class names and throw an error if a class fails to implement correctly?

gigabo replied:
Oops, your second comment didn't show up when I was typing my response.

Clashing names would be a problem no matter which way they're specified.

How about errors on:
- Invalid class name
- Name collision

That will make it easier to catch problems in the future.

gigabo replied:
Okay, I've got this queued up for a follow-up PR, since it works well as a standalone.

Also added an error on:
- A class that doesn't extend `Command`

Will submit after this lands.

rygine commented on src/Command.js:
```diff
@@ -166,6 +207,19 @@ export default class Command {
 
+  _legacyOptions() {
+    let opts = {};
+    if (this.name === "bootstrap" && (this.repository.asiniJson || {}).bootstrapConfig) {
```
i think this could be part of `repository.asiniJson`, always ensuring an object is returned.

gigabo replied:
Yeah, I wavered on that.  Wasn't sure if there would ever be a case where we cared whether it was even defined.  I guess probably not.  I'll clean this up.

rygine commented on src/Command.js:
```diff
@@ -166,6 +207,19 @@ export default class Command {
+    }
+    return opts;
+  }
+
```
what about something like this?

``` javascript
["bootstrap", "publish"].forEach((command) => {
  if (this.name === command && this.repository.asiniJson[`${command}Config`]) {
    logger.warn(`\`${command}Config.ignore\` is deprecated.  Use \`commands.${command}.ignore\`.`);
    opts.ignore = this.repository.asiniJson.[`${command}Config`].ignore;
  }
});
```

gigabo replied:
:metal:

gigabo committed:
Use objectAssign for node 0.10 & 0.12 compatibility

gigabo committed:
Factor out asiniJson default empty object

gigabo committed:
Slicker legacy option handling courtesy of @rygine

rygine commented on src/Command.js:
```diff
@@ -19,6 +19,47 @@ export default class Command {
+
+      // CLI flags always override everything.
+      this.flags,
+    );
```
i think the Flags section of the README need an entire section talking about precedence.

gigabo replied:
Oh, yeah... good call!

gigabo committed:
Just pass options to `buildPackageGraph`

Cleaner interface.

No need to send the whole command object through just for one method.

gigabo committed:
Add tests for legacy options

gigabo committed:
Add some explanation of the config sieve to the README

gigabo commented:
Alright, unless there are any objections I'm going to merge this in a few and get the follow up command validation PR up for review.

rygine reviewed (approved):
LGTM
